tensor_comprehensions/__init__.py

```diff
@@ -12,20 +12,372 @@
+        return autotune(
+            tc,
+            entry_point,
+            *inputs,
+            starting_options=starting_options,
+            tuner_config=tuner_config,
+            cache_filename=cache_filename,
+            load_from_cache=load_from_cache,
+            store_to_cache=store_to_cache)
+
+    return generate
+
+class TC(object):
+    def __init__(
+            self,
+            tc: str,
+            mapping_options_factory: (
+                Callable[[str, str, Iterable[torch.Tensor]], MappingOptions])
+    ):
+        self.tc = tc
+        self.mapping_options_factory = mapping_options_factory
+        self.compilation_cache = CompilationCache(self.tc)
+        # Make each TC def in the tc str a method of the TC object so we can:
+        #     T = tc.define("def add() ...")
+        #     T.add()
+        #
+        def make_closure(obj: TC, tc_def_name: str):
+            def fun(*inputs: torch.Tensor,
+                    outputs: Optional[Tuple[torch.Tensor]] = (),
+                    unchecked: Optional[bool] = False) -> List[torch.Tensor] :
+                return obj.__call__(
+                    tc_def_name, *inputs, outputs=outputs, unchecked=unchecked)
+
+            return fun
+
+        for tc_def in tclib.parse_defs(self.tc):
+            self.__setattr__(tc_def, make_closure(self, tc_def))
+
+    def __call__(
+            self,
+            entry_point: str,
+            *inputs: torch.Tensor,
+            outputs: Optional[Tuple[torch.Tensor]] = (),
+            unchecked: Optional[bool] = False) -> List[torch.Tensor]:
+        r"""
+        Runs the defined TC function on given inputs. If the TC function has not
+        yet been compiled for the particular input sizes it will be JIT-compiled
+        and memoized the first time. The MappingOptions used to compile the TC
+        function is determined by the TC object's  mapping_options_factory which
+        can be (re)set for a TC object by calling set_mapping_options_factory.
+
+        Args:
+            entry_point: the name of the TC def to be executed
+
+            inputs: tensors that TC takes as input arguments. The inputs must
+                be passed in the order of the definition of the TC function.
+
+            outputs: tensors the TC writes to in-place. If output tensors are
+                omitted, new outputs will be allocated and returned at each
+                call.
+
+            unchecked: run in low-latency unchecked mode where failsafe size
+                and stride checks are omitted.
+
+        Returns
+            List[torch.Tensor]: the tensors returned by the TC. If outputs are
+                not specified then new allocations take place. Otherwise the TC
+                writes in-place and returns the same tensors it received as
+                outputs.
+        """
+
+        # Locally scoped implicit compilation
+        def __implicit_compile__(tc_obj: TC,
```

Comment:
you don't need the `__name__` syntax here, no name mangling happens at the function scope